### PR TITLE
Allow any floating-point value as a 3D rendering scale option

### PR DESCRIPTION
**Pull request by Calinou**
Follow-up to https://github.com/godotengine/godot/pull/51870.

This allows for finer control over 3D rendering resolution. Supersampling can also be performed by setting a 3D rendering resolution above 1.0, which is useful for offline rendering or for very high-end GPUs.

With this change, it's possible to implement dynamic resolution scaling using a script. It's not ideal but seems to be usable enough, provided you disable temporal reprojection in volumetric fog (if you use volumetric fog). The sky will also flicker black when the render scale is changed for one frame, but we can figure that out in a later PR.

**Testing project:** [test_render_scale.zip](https://github.com/godotengine/godot/files/7072553/test_render_scale.zip)
Use ZQSD/Space/Shift to move, mouse wheel up/down or <kbd>F2</kbd>/<kbd>F1</kbd> to change the rendering scale. Press <kbd>X</kbd> to toggle the experimental dynamic resolution scaling (it targets ~200 FPS with the frametime variables, modify `Label.gd` to change this).

___

I haven't submitted this in the PR, but it appears that anamorphic resolution scaling (different factors for width and height) works fine with the above testing project. If desired, I can amend this pull request or open a new one after this PR is merged.

### CODE REVIEW

**Comment by Calinou** [comment]
Superseded by https://github.com/godotengine/godot/pull/52232. (Leaving this PR as reference if needed.)

**Comment by Calinou** [comment]
Reopening per https://github.com/godotengine/godot/pull/52232#issuecomment-908448190.

File: servers/rendering/renderer_viewport.cpp [hunk 1]
```diff
@@ -77,33 +77,17 @@ void RendererViewport::_configure_3d_render_buffers(Viewport *p_viewport) {
...
-				case RS::VIEWPORT_SCALE_3D_25_PERCENT: {
-					width = width >> 2;
-					height = height >> 2;
-				}; break;
-				default:
-					break;
-			}
+			// Clamp 3D rendering resolution to reasonable values supported on most hardware.
+			// This prevents freezing the engine or outright crashing on lower-end GPUs.
+			const int width = CLAMP(p_viewport->size.width * scale_3d, 1, 16384);
```
Comment (reduz) [inline servers/rendering/renderer_viewport.cpp:89]: You should definitely not be able to scale beyond 1.0, so I would clamp scale_3d to 1.0 instead.

**Review by reduz — CHANGES_REQUESTED** [review]

File: servers/rendering/renderer_viewport.cpp (hunk 1 above)
Comment (Calinou) [reply servers/rendering/renderer_viewport.cpp:89]: Scaling beyond 1.0 makes it possible to achiece supersampling in a way that's cheaper and easier than using a viewport with a larger size. While supersampling is not very practical for games, it's very useful for offline rendering, marketing screenshots, photo modes, etc.

**Review by reduz — APPROVED** [review]

**Pull request merged**